File: dataprivacy-labs-main/dataprivacy-labs-main/lab2/haskell/src/Examples.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}

{- |
Module      : Examples
Description : Example usage of the Differential Privacy DSL with Adult Census Dataset

This module demonstrates how to use the various components of the DSL
with real census data, showing practical differential privacy applications.
-}
module Examples where

import AdultDataset (
    Adult (..),
    Income (..),
    Occupation (..),
    loadAdultDataset,
 )
import Core (DPResult (..), Dataset (Dataset), Epsilon)
import qualified Data.Map as Map
import Measurements (dpCount)
import PrivacyTracker (Query, runQuery)
import Transformations (
    dpFilter,
    dpGroupBy,
    dpMap,
    dpPartition,
    dpUnion,
 )

-- ============================================================================
-- Differential Privacy Examples with Adult Census Dataset
-- ============================================================================

-- ============================================================================
-- Basic Examples with Adult Census Dataset
-- ============================================================================

-- | Basic example: counting high earners with multiple epsilon values in one session
basicCountExample :: Dataset 1 Adult -> IO ()
basicCountExample adultData = do
    putStrLn "=== High Earners Count Example ==="

    let Dataset adults = adultData
        trueHighEarnerCount = length [a | a <- adults, income a == HighIncome]
    putStrLn $ "True high earners count: " ++ show trueHighEarnerCount

    -- Run all three queries in a single session with budget composition
    let epsilons = [0.1, 1.0, 5.0]
        totalBudget = sum epsilons
    results <- runQuery (multiEpsilonCountQuery adultData) totalBudget

    putStrLn "DP Count with different epsilon values (single session):"
    mapM_
        ( \(eps, result) ->
            putStrLn $ "  ε=" ++ show eps ++ ": " ++ show (dpValue result)
        )
        (zip epsilons results)

-- | Run multiple count queries with different epsilon values in one session
multiEpsilonCountQuery :: Dataset 1 Adult -> Query [DPResult]
multiEpsilonCountQuery dataset = do
    highEarners <- dpFilter (\person -> income person == HighIncome) dataset

    -- Run three count queries with different epsilon allocations
    result1 <- dpCount highEarners 0.1
    result2 <- dpCount highEarners 0.5
    result3 <- dpCount highEarners 1.0

    return [result1, result2, result3]

-- | Example showing union and stability amplification with overlapping groups
unionStabilityExample :: Dataset 1 Adult -> IO ()
unionStabilityExample adultData = do
    putStrLn "=== Union with Overlapping Groups Example ==="

    let Dataset adults = adultData
        -- Two overlapping groups: high earners and white-collar workers
        highEarnersCount = length [a | a <- adults, income a == HighIncome]
        whiteCollarCount =
            length
                [ a
                | a <- adults
                , occupation a `elem` [ExecManagerial, ProfSpecialty, TechSupport, AdmClerical]
                ]
        -- Count people in both groups (overlap)
        overlapCount = highEarnersCount + whiteCollarCount

    putStrLn $ "True high earners count: " ++ show highEarnersCount
    putStrLn $ "True white-collar workers count: " ++ show whiteCollarCount
    putStrLn $ "Expected union count: " ++ show overlapCount

    -- Run with different epsilon values in single session
    let epsilons = [0.1, 1.0, 5.0]
        totalBudget = sum epsilons
    results <- runQuery (multiEpsilonUnionQuery adultData) totalBudget

    putStrLn "DP Union count with different epsilon values (single session):"
    mapM_
        ( \(eps, result) ->
            putStrLn $ "  ε=" ++ show eps ++ ": " ++ show (dpValue result)
        )
        (zip epsilons results)

-- | Run multiple union queries with different epsilon values in one session
multiEpsilonUnionQuery :: Dataset 1 Adult -> Query [DPResult]
multiEpsilonUnionQuery dataset = do
    -- Group 1: High earners (income > 50K)
    highEarners <- dpFilter (\person -> income person == HighIncome) dataset

    -- Group 2: White-collar workers (overlaps with high earners)
    whiteCollar <-
        dpFilter
            ( \person ->
                occupation person `elem` [ExecManagerial, ProfSpecialty, TechSupport, AdmClerical]
            )
            dataset

    -- Union creates stability amplification since some people appear in both groups
    combined <- dpUnion highEarners whiteCollar

    -- Run three union count queries with different epsilon allocations
    result1 <- dpCount combined 0.1
    result2 <- dpCount combined 1.0
    result3 <- dpCount combined 5.0

    return [result1, result2, result3]

-- ============================================================================
-- GroupBy Example with Occupations
-- ============================================================================

-- | Example using groupBy to analyze occupations with high work hours
occupationAnalysisExample :: Dataset 1 Adult -> IO ()
occupationAnalysisExample adultData = do
    putStrLn "=== Occupation GroupBy Example ==="

    let Dataset adults = adultData
        -- Calculate true average hours per occupation using Map.fromListWith
        occupationGroups = Map.toList $ Map.fromListWith (++) [(occupation a, [a]) | a <- adults]
        occupationAvgs =
            [ (occ, if null people then 0 else fromIntegral (sum [hoursPerWeek p | p <- people]) / fromIntegral (length people))
            | (occ, people) <- occupationGroups
            ]
        trueHighAvgOccupations = length [(occ, avg) | (occ, avg) <- occupationAvgs, avg > 40]

    putStrLn "True average hours per occupation:"
    mapM_ (\(occ, avg) -> putStrLn $ "  " ++ show occ ++ ": " ++ show @Double avg) occupationAvgs
    putStrLn $ "True count of occupations with avg hours > 40: " ++ show trueHighAvgOccupations

    -- Run with different epsilon values in single session
    let epsilons = [1.0, 1.5, 2.0, 10]
        totalBudget = sum epsilons
    results <- runQuery (multiEpsilonOccupationQuery adultData) totalBudget

    putStrLn "DP count of occupations with avg hours > 40 (single session):"
    mapM_
        ( \(eps, result) ->
            putStrLn $ "  ε=" ++ show eps ++ ": " ++ show (dpValue result)
        )
        (zip epsilons results)

-- | Run multiple occupation analysis queries with different epsilon values in one session
multiEpsilonOccupationQuery :: Dataset 1 Adult -> Query [DPResult]
multiEpsilonOccupationQuery dataset = do
    -- Group by occupation (this operation is shared across all queries)
    occGroups <- dpGroupBy occupation dataset

    -- Calculate average hours per occupation (shared computation)
    avgHours <-
        dpMap
            ( \(_occ, people) ->
                let totalHours = sum [hoursPerWeek p | p <- people]
                    count = length people
                 in if count > 0 then fromIntegral @Int @Double totalHours / fromIntegral count else 0
            )
            occGroups

    -- Filter to occupations with high average hours (shared)
    hardWorkingOccs <- dpFilter (> 40) avgHours

    -- Run three count queries with different epsilon allocations
    result1 <- dpCount hardWorkingOccs 1.0
    result2 <- dpCount hardWorkingOccs 1.5
    result3 <- dpCount hardWorkingOccs 2.0
    result4 <- dpCount hardWorkingOccs 10

    return [result1, result2, result3, result4]

-- ============================================================================
-- Partition Example with Age Groups
-- ============================================================================

-- | Demonstrate partition with age-based analysis
partitionExample :: Dataset 1 Adult -> IO ()
partitionExample adultData = do
    putStrLn "=== Age Group Partition Example ==="

    let Dataset adults = adultData
        youngCount = length [a | a <- adults, age a < 30, income a == HighIncome]
        middleCount = length [a | a <- adults, age a >= 30 && age a < 50, income a == HighIncome]
        seniorCount = length [a | a <- adults, age a >= 50, income a == HighIncome]

    putStrLn "True high earner counts by age group:"
    putStrLn $ "  Young (<30): " ++ show youngCount
    putStrLn $ "  Middle (30-49): " ++ show middleCount
    putStrLn $ "  Senior (50+): " ++ show seniorCount

    -- Run with different epsilon values in single session
    let epsilons = [0.1, 1.0, 10.0]
        totalBudget = sum epsilons
    results <- runQuery (multiEpsilonPartitionQuery adultData) totalBudget

    putStrLn "DP high earner counts by age group with different epsilon values (single session):"
    mapM_
        ( \(eps, ageGroupResults) -> do
            putStrLn $ "  ε=" ++ show eps ++ ":"
            mapM_
                ( \(ageGroup, result) ->
                    putStrLn $ "    " ++ ageGroup ++ ": " ++ show (dpValue result)
                )
                ageGroupResults
        )
        (zip epsilons results)

-- | Run multiple partition queries with different epsilon values in one session
multiEpsilonPartitionQuery :: Dataset 1 Adult -> Query [[(String, DPResult)]]
multiEpsilonPartitionQuery dataset = do
    -- Run three partition queries with different epsilon allocations
    result1 <- ageGroupPartitionQuery dataset 0.1
    result2 <- ageGroupPartitionQuery dataset 1.0
    result3 <- ageGroupPartitionQuery dataset 10.0

    return [result1, result2, result3]

ageGroupPartitionQuery :: Dataset 1 Adult -> Epsilon -> Query [(String, DPResult)]
ageGroupPartitionQuery dataset eps = do
    dpPartition eps ageGroups classifyAge dataset countHighEarners
  where
    ageGroups = ["Young", "Middle", "Senior"]
    classifyAge person
        | age person < 30 = "Young"
        | age person < 50 = "Middle"
        | otherwise = "Senior"
    countHighEarners _ageGroup ageGroupData = do
        highEarners <- dpFilter (\person -> income person == HighIncome) ageGroupData
        dpCount highEarners eps

-- ============================================================================
-- Main Function
-- ============================================================================

-- | Run all examples with the Adult Census dataset
main :: IO ()
main = do
    putStrLn "Differential Privacy DSL - Adult Census Dataset Examples"
    putStrLn (replicate 60 '=')

    putStrLn "Loading Adult Census dataset..."
    adultData <- loadAdultDataset

    let Dataset adults = adultData
    putStrLn $ "Loaded " ++ show (length adults) ++ " records"
    putStrLn ""

    -- Run all examples with the loaded dataset
    basicCountExample adultData
    putStrLn ""
    --
    unionStabilityExample adultData
    putStrLn ""

    occupationAnalysisExample adultData
    putStrLn ""
    --
    partitionExample adultData
    putStrLn ""

    putStrLn "All examples completed!"

File: dataprivacy-labs-main/dataprivacy-labs-main/lab2/python/examples.py
"""
Example usage of the Differential Privacy DSL with Adult Census Dataset

This module demonstrates how to use the DP_DSL class with real census data,
showing practical differential privacy applications.
"""

from typing import List, Tuple, Any
from dp import DP_DSL, Dataset, create_dataset, Adult, Income, Occupation, load_adult_dataset


def basic_count_example(adult_data: Dataset[Adult]) -> None:
    """Basic example: counting high earners with different epsilon values"""
    print("=== High Earners Count Example ===")

    # Calculate true high earner count
    true_high_earner_count = len([a for a in adult_data if a.income == Income.HIGH_INCOME])
    print(f"True high earners count: {true_high_earner_count}")

    # Create DSL instance with total budget for all queries
    dsl = DP_DSL(budget=6.1)  # 0.1 + 1.0 + 5.0

    # Filter to high earners (budget-free transformation)
    high_earners = dsl.filter(lambda person: person.income == Income.HIGH_INCOME, adult_data)

    # Run three count queries with different epsilon values
    print("DP Count with different epsilon values:")

    result1 = dsl.count(high_earners, epsilon=0.1)
    print(f"  ε=0.1: {result1.dp_value:.2f}")
    print(f"  Budget status: {dsl}")

    result2 = dsl.count(high_earners, epsilon=1.0)
    print(f"  ε=1.0: {result2.dp_value:.2f}")
    print(f"  Budget status: {dsl}")

    result3 = dsl.count(high_earners, epsilon=5.0)
    print(f"  ε=5.0: {result3.dp_value:.2f}")
    print(f"  Final budget status: {dsl}")


def union_stability_example(adult_data: Dataset[Adult]) -> None:
    """Example showing union and stability amplification"""
    print("=== Union with Overlapping Groups Example ===")

    # Calculate true counts
    high_earners_count = len([a for a in adult_data if a.income == Income.HIGH_INCOME])

    white_collar_occupations = {
        Occupation.EXEC_MANAGERIAL,
        Occupation.PROF_SPECIALTY,
        Occupation.TECH_SUPPORT,
        Occupation.ADM_CLERICAL
    }
    white_collar_count = len([
        a for a in adult_data
        if a.occupation in white_collar_occupations
    ])

    overlap_count = high_earners_count + white_collar_count

    print(f"True high earners count: {high_earners_count}")
    print(f"True white-collar workers count: {white_collar_count}")
    print(f"Expected union count: {overlap_count}")

    # Create DSL instance
    dsl = DP_DSL(budget=6.1)

    # Create groups (budget-free transformations)
    high_earners = dsl.filter(lambda person: person.income == Income.HIGH_INCOME, adult_data)
    white_collar = dsl.filter(
        lambda person: person.occupation in white_collar_occupations,
        adult_data
    )

    # Union creates stability amplification (budget-free)
    combined = dsl.union(high_earners, white_collar)
    print(f"Combined dataset stability: {combined.stability}")

    # Run union count queries with different epsilon values
    print("DP Union count with different epsilon values:")

    result1 = dsl.count(combined, epsilon=0.1)
    print(f"  ε=0.1: {result1.dp_value:.2f}")

    result2 = dsl.count(combined, epsilon=1.0)
    print(f"  ε=1.0: {result2.dp_value:.2f}")

    result3 = dsl.count(combined, epsilon=5.0)
    print(f"  ε=5.0: {result3.dp_value:.2f}")

    print(f"Final budget status: {dsl}")


def occupation_analysis_example(adult_data: Dataset[Adult]) -> None:
    """Example using groupBy to analyze occupations"""
    print("=== Occupation GroupBy Example ===")

    # Calculate true average hours per occupation
    from collections import defaultdict
    occupation_groups = defaultdict(list)
    for person in adult_data:
        occupation_groups[person.occupation].append(person)

    occupation_avgs = []
    for occ, people in occupation_groups.items():
        if people:
            avg_hours = sum(p.hours_per_week for p in people) / len(people)
            occupation_avgs.append((occ, avg_hours))

    true_high_avg_occupations = len([(occ, avg) for occ, avg in occupation_avgs if avg > 40])

    print("True average hours per occupation:")
    for occ, avg in occupation_avgs[:10]:  # Show first 10
        print(f"  {occ.name}: {avg:.2f}")
    print(f"True count of occupations with avg hours > 40: {true_high_avg_occupations}")

    # Create DSL instance
    dsl = DP_DSL(budget=15.5)  # 1.0 + 1.5 + 2.0 + 11.0

    # Group by occupation (budget-free, but doubles stability)
    occ_groups = dsl.group_by(lambda person: person.occupation, adult_data)
    print(f"Grouped dataset stability: {occ_groups.stability}")

    # Calculate average hours per occupation (budget-free)
    def calc_avg_hours(occ_people_tuple):
        occ, people = occ_people_tuple
        if people:
            total_hours = sum(p.hours_per_week for p in people)
            return total_hours / len(people)
        return 0.0

    avg_hours = dsl.map(calc_avg_hours, occ_groups)

    # Filter to occupations with high average hours (budget-free)
    hard_working_occs = dsl.filter(lambda avg: avg > 40, avg_hours)

    # Run count queries with different epsilon values
    print("DP count of occupations with avg hours > 40:")

    result1 = dsl.count(hard_working_occs, epsilon=1.0)
    print(f"  ε=1.0: {result1.dp_value:.2f}")

    result2 = dsl.count(hard_working_occs, epsilon=1.5)
    print(f"  ε=1.5: {result2.dp_value:.2f}")

    result3 = dsl.count(hard_working_occs, epsilon=2.0)
    print(f"  ε=2.0: {result3.dp_value:.2f}")

    result4 = dsl.count(hard_working_occs, epsilon=11.0)
    print(f"  ε=11.0: {result4.dp_value:.2f}")

    print(f"Final budget status: {dsl}")


def partition_example(adult_data: Dataset[Adult]) -> None:
    """Demonstrate partition with age-based analysis"""
    print("=== Age Group Partition Example ===")

    # Calculate true high earner counts by age group
    young_count = len([a for a in adult_data
                      if a.age < 30 and a.income == Income.HIGH_INCOME])
    middle_count = len([a for a in adult_data
                       if 30 <= a.age < 50 and a.income == Income.HIGH_INCOME])
    senior_count = len([a for a in adult_data
                       if a.age >= 50 and a.income == Income.HIGH_INCOME])

    print("True high earner counts by age group:")
    print(f"  Young (<30): {young_count}")
    print(f"  Middle (30-49): {middle_count}")
    print(f"  Senior (50+): {senior_count}")

    # Create DSL instance for each partition query
    results = []

    for epsilon in [0.1, 1.0, 10.0]:
        # Create fresh DSL instance for each epsilon test
        dsl = DP_DSL(budget=epsilon + 0.1)  # Small buffer for partition overhead

        def classify_age(person: Adult) -> str:
            if person.age < 30:
                return "Young"
            elif person.age < 50:
                return "Middle"
            else:
                return "Senior"

        def count_high_earners(age_group: str, age_group_data: Dataset[Adult]) -> Any:
            high_earners = dsl.filter(
                lambda person: person.income == Income.HIGH_INCOME,
                age_group_data
            )
            return dsl.count(high_earners, epsilon=epsilon)

        # Use partition (which consumes epsilon for the operation)
        partition_results = dsl.partition(
            keys=["Young", "Middle", "Senior"],
            classifier=classify_age,
            dataset=adult_data,
            computation=count_high_earners,
            epsilon=epsilon
        )

        results.append((epsilon, partition_results))

    print("DP high earner counts by age group with different epsilon values:")
    for epsilon, age_group_results in results:
        print(f"  ε={epsilon}:")
        for age_group, result in age_group_results:
            print(f"    {age_group}: {result.dp_value:.2f}")


def budget_management_example() -> None:
    """Example showing advanced budget management features"""
    print("=== Budget Management Example ===")

    # Create a small test dataset
    test_data = create_dataset([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

    # Create DSL instance
    dsl = DP_DSL(budget=3.0)
    print(f"Initial: {dsl}")

    # Apply transformations (budget-free)
    filtered = dsl.filter(lambda x: x > 5, test_data)
    print(f"After filter: {dsl}")

    # First measurement
    result1 = dsl.count(filtered, epsilon=1.0)
    print(f"After first count (ε=1.0): {dsl}")
    print(f"Result 1: {result1.dp_value:.2f}")

    # Check budget status
    status = dsl.budget_status()
    print(f"Budget status: {status}")

    # Second measurement
    doubled = dsl.map(lambda x: x * 2.0, filtered)
    result2 = dsl.sum(doubled, epsilon=1.5, lower_bound=0.0, upper_bound=20.0)
    print(f"After sum (ε=1.5): {dsl}")
    print(f"Result 2: {result2.dp_value:.2f}")

    # Try to exceed budget (should fail)
    print("\nTrying to exceed budget...")
    try:
        result3 = dsl.count(filtered, epsilon=1.0)  # Would exceed budget
        print("ERROR: Should have failed!")
    except Exception as e:
        print(f"Correctly caught error: {e}")

    print(f"Final: {dsl}")


def main() -> None:
    """Run all examples with the Adult Census dataset"""
    print("Differential Privacy DSL - Examples")
    print("=" * 50)

    print("Loading Adult Census dataset...")
    adult_data = load_adult_dataset("adult.csv")

    print(f"Loaded {len(adult_data)} records")
    print()

    # Run all examples with the loaded dataset
    basic_count_example(adult_data)
    print()

    union_stability_example(adult_data)
    print()

    occupation_analysis_example(adult_data)
    print()

    partition_example(adult_data)
    print()

    budget_management_example()
    print()

    print("All examples completed!")


if __name__ == "__main__":
    main()